packages/projection/README.md

```diff
@@ -1,20 +1,100 @@
+2. Events include some additional properties: `{eraSumaries, genesisParameters, epochNo, crossEpochBoundary}`.
+3. `RollBackward` events include block data (instead of just specifying the rollback point), and are emitted once for **each** rolled back block.
+
+#### ExtraProps (Generic Parameter)
+
+Source observable can be piped through a series of RxJS operators, which may add some properties to the event. In a nutshell, `ProjectionEvent<T> = ProjectionEvent<{}> & T`. For more detail see [Mappers](#mappers).
+
+### StabilityWindowBuffer
+
+Since `ProjectionEvent{eventType=RollBackward}` includes block data, we must store some number (at least `k`, which is the "security parameter") of latest blocks so that if a fork/rollback happens, we can call `StabilityWindowBuffer.getBlock(hash)` in order to build a valid `ProjectionEvent`.
+
+### ObservableCardanoNode
+
+This is our interface to Chain Sync and Local State Query protocols of Cardano node. The only implementation available in `cardano-js-sdk` is `OgmiosObservableCardanoNode`.
+
+## Utilities
+
+### Bootstrap.fromCardanoNode
+
+Creates a projection source (`Observable<ProjectionEvent>`) from:
+- [ObservableCardanoNode](#observablecardanonode)
+- [StabilityWindowBuffer](#stabilitywindowbuffer)
+- `Observable<TipOrOrigin>` that emits block header of last processed event (local tip). This is used to find intersection point.
+
+```mermaid
+---
+title: Bootstrap Algorithm
+---
+flowchart LR
+    subscribe[Subscribe] --> find-intersect{Find node\nintersection\nwith local tip}
+    find-intersect --> |Found| start[Start ChainSync\nfrom intersection]
+    start --> get-event{Get event\nfrom node,\ncheck type}
+    get-event --> |RollBackward| get-block[Get block from StabilityWindowBuffer]
+    get-event --> |RollForward| hydrate
+    start --> hydrate[Hydrate event\nwith additional\npropertes]
+    hydrate --> emit[Emit\nProjectionEvent]
+    find-intersect --> |Not Found| get-block[Get block from\nStabilityWindowBuffer]
+    get-block --> rollback[Create\nRollBackward\nevent]
+    rollback --> hydrate
+    rollback --> wait[Wait for\nlocal tip change]
+    wait --> is-rollback-point{Is rollback point?}
+    is-rollback-point --> |No| get-block
+    is-rollback-point --> |Yes| started-sync{Already started\nChain Sync?}
+    started-sync --> |Yes| get-event
+    started-sync --> |No| find-intersect
+```
+
+### Mappers
+
+Projections are typically 
+1. Interpreting the block (e.g. extracting some relevant properties from all transactions). `Mappers` is a namespace that contains a collection of RxJS operators that process the ProjectionEvent (block) and emit a new event object that has some additional properties.
+2. Doing something with it (e.g. store into the database). 
+
+Here's an example:
+
+```typescript
+// source$: Observable<ProjectionEvent<{}>>
+source$.pipe(
+  // policyId is a parameter that should be provided
+  // by handle issuer, e.g. ADA Handle
+  Mappers.withHandles({ policyId }),
```

Comment:
I am tempted to think that "withHandles" also does filtering. But I understand that it does not. If there are no handles for `policyId`, it will still emit the ProjectionEvent, with emtpy `event.handles`.
I my understanding correct?

Reply:
Yes, it would still emit an event with empty handles. Filtering is a good idea and it would improve performance quite a lot, but it should be implemented on the application level (cardano-services), because otherwise we'd lose ability to project multiple things